File: ios/ExpoNativeCompactTabsModule.swift
import ExpoModulesCore

struct ExpoNativeCompactTabItem: Record {
  @Field var key: String = ""
  @Field var label: String = ""
  @Field var imageUri: String = ""
  @Field var animationFrameUris: [String] = []
  @Field var accessibilityLabel: String?
}

public final class ExpoNativeCompactTabsModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoNativeCompactTabs")

    View(ExpoNativeCompactTabsView.self) {
      Events("onTabSelected")

      Prop("items") { (view: ExpoNativeCompactTabsView, items: [ExpoNativeCompactTabItem]) in
        view.setItems(items)
      }

      Prop("selectedIndex") { (view: ExpoNativeCompactTabsView, index: Int) in
        view.setSelectedIndex(index)
      }

      Prop("compact") { (view: ExpoNativeCompactTabsView, compact: Bool) in
        view.setCompact(compact)
      }

      Prop("tintColor") { (view: ExpoNativeCompactTabsView, color: UIColor?) in
        view.tabTintColor = color
      }

      Prop("inactiveTintColor") { (view: ExpoNativeCompactTabsView, color: UIColor?) in
        view.inactiveTabTintColor = color
      }

      Prop("expandedHorizontalInset") { (view: ExpoNativeCompactTabsView, inset: CGFloat?) in
        view.expandedHorizontalInset = inset ?? 12
      }

      Prop("compactWidth") { (view: ExpoNativeCompactTabsView, width: CGFloat?) in
        view.compactWidth = width ?? 352
      }

      Prop("expandedHeight") { (view: ExpoNativeCompactTabsView, height: CGFloat?) in
        view.expandedHeight = height ?? 78
      }

      Prop("compactHeight") { (view: ExpoNativeCompactTabsView, height: CGFloat?) in
        view.compactHeight = height ?? 70
      }

      Prop("animationFrameDuration") { (view: ExpoNativeCompactTabsView, duration: Double?) in
        view.animationFrameDuration = duration ?? 0.034
      }
    }
  }
}

File: android/src/main/java/expo/modules/nativecompacttabs/ExpoNativeCompactTabsModule.kt
package expo.modules.nativecompacttabs

import android.graphics.Color
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import expo.modules.kotlin.types.OptimizedRecord

@OptimizedRecord
class ExpoNativeCompactTabItem : Record {
  @Field
  var key: String = ""

  @Field
  var label: String = ""

  @Field
  var imageUri: String = ""

  @Field
  var animationFrameUris: List<String> = emptyList()

  @Field
  var accessibilityLabel: String? = null
}

@Suppress("unused")
class ExpoNativeCompactTabsModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoNativeCompactTabs")

    View(ExpoNativeCompactTabsView::class) {
      Events("onTabSelected")

      Prop("items") { view: ExpoNativeCompactTabsView, items: List<ExpoNativeCompactTabItem> ->
        view.setItems(items)
      }

      Prop("selectedIndex") { view: ExpoNativeCompactTabsView, index: Int ->
        view.setSelectedIndex(index)
      }

      Prop("compact") { view: ExpoNativeCompactTabsView, compact: Boolean ->
        view.setCompact(compact)
      }

      Prop("tintColor") { view: ExpoNativeCompactTabsView, color: Color? ->
        view.setTintColor(color)
      }

      Prop("inactiveTintColor") { view: ExpoNativeCompactTabsView, color: Color? ->
        view.setInactiveTintColor(color)
      }

      Prop("expandedHorizontalInset") { view: ExpoNativeCompactTabsView, inset: Double? ->
        view.expandedHorizontalInset = inset?.toFloat() ?: 12f
      }

      Prop("compactWidth") { view: ExpoNativeCompactTabsView, width: Double? ->
        view.compactWidth = width?.toFloat() ?: 352f
      }

      Prop("expandedHeight") { view: ExpoNativeCompactTabsView, height: Double? ->
        view.expandedHeight = height?.toFloat() ?: 78f
      }

      Prop("compactHeight") { view: ExpoNativeCompactTabsView, height: Double? ->
        view.compactHeight = height?.toFloat() ?: 70f
      }

      Prop("animationFrameDuration") { view: ExpoNativeCompactTabsView, duration: Double? ->
        view.animationFrameDuration = duration ?: 0.034
      }

      Prop("selectionDragEnabled") { view: ExpoNativeCompactTabsView, enabled: Boolean? ->
        view.selectionDragEnabled = enabled ?: true
      }
    }
  }
}
